python/shotgun_data/background_task_manager.py

```diff
@@ -0,0 +1,832 @@
+                tasks_to_stop.append(task)
+        del self._group_task_map[group]
+        self._stop_tasks(tasks_to_stop, stop_upstream, stop_downstream)
+
+        self._log(" > Task group %s stopped!" % group)
+
+    def stop_all_tasks(self):
+        """
+        Stop all currently queued or running tasks.  If any tasks are already running then the will 
+        complete but their completion/failure signals will be ignored.
+        """
+        self._log("Stopping all tasks...")
+
+        # we just need to clear all the lookups:
+        self._running_tasks = {}
+        self._pending_tasks_by_priority = {}
+        self._tasks_by_id = {}
+        self._group_task_map = {}
+        self._upstream_task_map = {}
+        self._downstream_task_map = {}
+
+        self._log(" > All tasks stopped!")
+
+    def _stop_tasks(self, tasks_to_stop, stop_upstream, stop_downstream):
+        """
+        Stop the specified list of tasks
+
+        :param tasks_to_stop:   A list of tasks to stop
+        :param stop_upstream:   If true then all upstream tasks will also be stopped
+        :param stop_downstream: If true then all downstream tasks will also be stopped
+        """
+        if not tasks_to_stop:
+            return
+
+        # copy the task list as we'll be modifying it:
+        tasks_to_stop = list(tasks_to_stop)
+        # and make sure we only stop each task once!!
+        stopped_task_ids = set([task.uid for task in tasks_to_stop])
+
+        while tasks_to_stop:
+            task_to_stop = tasks_to_stop.pop(0)
+
+            # get the up & downstream tasks to also stop depending on the flags:
+            if stop_upstream and task_to_stop.uid in self._upstream_task_map:
+                for us_task_id in self._upstream_task_map[task_to_stop.uid]:
+                    us_task = self._tasks_by_id.get(us_task_id)
+                    if not us_task or us_task.uid in stopped_task_ids:
+                        # no task or already found
+                        continue
+
+                    tasks_to_stop.append(us_task)
+                    stopped_task_ids.add(us_task_id)
+
+            if stop_downstream and task_to_stop.uid in self._downstream_task_map:
+                for ds_task_id in self._downstream_task_map[task_to_stop.uid]:
+                    ds_task = self._tasks_by_id.get(ds_task_id)
+                    if not ds_task or ds_task.uid in stopped_task_ids:
+                        # no task or already found
+                        continue
+
+                    tasks_to_stop.append(ds_task)
+                    stopped_task_ids.add(ds_task_id)
+
+            # remove the task:
+            self._remove_task(task_to_stop)
+
+    def _get_worker_thread(self):
+        """
+        Get a worker thread to use.  
+
+        :returns:   An available worker thread if there is one, a new thread if needed or None if the thread
+                    limit has been reached.
+        """
+        if self._available_threads:
```

Comment:
All task processing is handled on the main thread as all task completion/failure signals are processed there but I will update the comments to be clear that the task manager itself is not thread safe.